### PR TITLE
Single websocket for queries

### PR DESCRIPTION
**Pull request by alexmatson-da**

### CODE REVIEW

File: ui/src/components/Exchange/ExchangeParticipants.tsx [hunk 1]
```diff
@@ -1,33 +1,28 @@
...
-import PageSection from '../common/PageSection'
-import Page from '../common/Page'
+import { useContractQuery } from '../../websocket/queryStream'
 
 import { depositSummary } from '../common/utils'
 import StripedTable from '../common/StripedTable'
+import PageSection from '../common/PageSection'
+import Page from '../common/Page'
+
+// import InviteParticipant from './InviteParticipant'
```
Comment (Sarahb-da) [inline ui/src/components/Exchange/ExchangeParticipants.tsx:15]: comment?

File: ui/src/components/Investor/Investor.tsx [hunk 2]
```diff
@@ -195,9 +187,11 @@ const Investor: React.FC<Props> = ({ onLogout }) => {
...
     </Switch>
 
-    return registeredInvestor.loading
-         ? loadingScreen
-         : registeredInvestor.contracts.length === 0 ? inviteScreen : investorScreen
+    return registeredInvestor.length === 0 ? inviteScreen : investorScreen
+
+    // return registeredInvestor.loading
+    //      ? loadingScreen
+    //      : registeredInvestor.length === 0 ? inviteScreen : investorScreen
```
Comment (Sarahb-da) [inline ui/src/components/Investor/Investor.tsx:194]: comment

File: ui/src/websocket/queryStream.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        ...queryStream,
+        streamErrors,
+      }))
+    }
+  }, [streamErrors, queryStream.streamErrors]);
+
+  return React.createElement(QueryStreamContext.Provider, { value: queryStream }, children);
+}
+
+// export function useContractQuery
```
Comment (Sarahb-da) [inline ui/src/websocket/queryStream.ts:133]: commeng

**Review by Sarahb-da — APPROVED** [review]
looks good

File: ui/src/components/Exchange/ExchangeParticipants.tsx (hunk 1 above)
Comment (alexmatson-da) [reply ui/src/components/Exchange/ExchangeParticipants.tsx:15]: oops, ty!

File: ui/src/websocket/queryStream.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+export function useContractQuery<T extends object, K = unknown, I extends string = string>(template: Template<T,K,I>, asPublic?: boolean) {
```
Comment (alexgraham-da) [inline ui/src/websocket/queryStream.ts:134]: Would it be possible to have a second function called `useContractQueryAsPublic` that just calls this one passing `true` into `asPublic` or would that cause problems with `useEffect`?

**Review by alexgraham-da — APPROVED** [review]
This is really great! Just one comment/question above.

Did this end up helping/solving the dropping websockets issue Eric was having?

File: ui/src/websocket/queryStream.ts (hunk 4 above)
Comment (alexmatson-da) [reply ui/src/websocket/queryStream.ts:134]: yep, I think that will work, but I'm not sure I see the benefit to having separate functions if the underlying implementation manages both of the public/private views anyway

**Comment by alexmatson-da** [comment]
> Did this end up helping/solving the dropping websockets issue Eric was having?

It's difficult to know for sure because of how tricky it's been to repro the problem, so only time will truly tell. I have some hopes because the reconnect logic is clearer and more pronounced. Strategically, if it doesn't pan out, I still see this work as beneficial because I'm planning to try to reuse some of it when opening websockets to exberry!

File: ui/src/websocket/queryStream.ts (hunk 4 above)
Comment (alexgraham-da) [reply ui/src/websocket/queryStream.ts:134]: yeah - the idea was to match `useStreamQueryAsPublic` and not have to worry about importing/passing a const around, but it's really personal preference as to which which you find cleaner, so I'm fine with either.

**Comment by alexgraham-da** [comment]
> > Did this end up helping/solving the dropping websockets issue Eric was having?
> 
> It's difficult to know for sure because of how tricky it's been to repro the problem, so only time will truly tell. I have some hopes because the reconnect logic is clearer and more pronounced. Strategically, if it doesn't pan out, I still see this work as beneficial because I'm planning to try to reuse some of it when opening websockets to exberry!

nice, sounds good!

**Pull request merged**